### ISSUE TITLE
`{% return %}` tag broken on Craft 5.4 / Twig 3.12?

### ISSUE THREAD

**Issue by rungta** [marionnewlevant/craft-twig_perversion#26]
The following seems to be failing in Craft 5.4.1:

```
{% macro returnTest() %}
  Hello
  {% return [1, 2, 3] %}
{% endmacro %}

{% dd returnTest() %}
```

Produces _Hello_ instead of an array.

**Comment by Kristomen** [comment]
I confirm that there is an issue after the update.
The result is like this  instead of an array
Twig\Markup {#1379 ▼
  -content: "                    "
  -charset: "UTF-8"
}

**Comment by knufri** [comment]
I encountered the same issue after upgraded to Craft 4.12.1 (Twig v3.12.0)! Any updates on this? On Craft 4.11.2 it is working as expected.

**Comment by mudechs** [comment]
same problem here after update to craft cms 4.12.1

**Comment by gmhenderson** [comment]
I can confirm the same issue. Our sites rely heavily on the `{% return %}` tag, so I'm very interested in a solution!

Previously, a simple macro like:
```twig
{% macro testMacro() %}
  {% return "test" %}
{% endmacro %}
```

would compile to:

```php

// ...

$blocks = [];

ob_start();
try {
    craft\helpers\Template::beginProfile("macro", "testMacro");
    // line 2
    echo "  ";
    return "test";
    craft\helpers\Template::endProfile("macro", "testMacro");

    return ('' === $tmp = ob_get_contents()) ? '' : new Markup($tmp, $this->env->getCharset());
} finally {
    ob_end_clean();
}
```

With Craft 5.4.1 it now compiles to:

```php
$blocks = [];

return ('' === $tmp = \Twig\Extension\CoreExtension::captureOutput((function () use (&$context, $macros, $blocks) {
    craft\helpers\Template::beginProfile("macro", "testMacro");
    // line 2
    yield "  ";
    return "test";
    craft\helpers\Template::endProfile("macro", "testMacro");
    return; yield '';
})())) ? '' : new Markup($tmp, $this->env->getCharset());
```

I don't know Twig internals well enough to make sense of this and what could be done about it. Any help would be greatly appreciated!

**Comment by marionnewlevant** [comment]
looking into it now...
See https://github.com/twigphp/Twig/issues/4146 https://github.com/craftcms/cms/discussions/15568
twig  5.4.0 updated twig to 3.12 (System changes)

**Comment by marionnewlevant** [comment]
I am not at all sure this is salvageable. Previously, a twig macro compiled into a more or less straightforward function, and it was more or less straightforward to return from that function.
In twig 3.12, a twig macro is much less straightforward.
I will see what I can do, but twig perversion is so named because it does things that twig is not required to support.

**Comment by gmhenderson** [comment]
@marionnewlevant thank you for taking a look. Unfortunately I came to the same conclusion. Now that Twig has moved to using generator functions with `yield` instead of `echo`, simply injecting an early `return` statement will no longer work, and I couldn't see a straightforward way around that. It's a sad day; I really loved all the "perverse" things I was able to do with macros using `return` 😭

**Comment by brandonkelly** [comment]
Just PR’d a fix for this: #27

**Comment by gmhenderson** [comment]
@brandonkelly Wow! Excellent! Thank you and very clever.

**Comment by rungta** [comment]
Hi @marionnewlevant, curious about your plans for the plugin. Is there a release planned with the #27 PR for the `{% return %}` tag, or will there be no more updates since the plugin has been marked as “no longer maintained”?

![image](https://github.com/user-attachments/assets/19a9f912-d3f3-40fb-9fea-e531b741eb7d)

**Comment by anita-chouhan** [comment]
Hello @marionnewlevant, any update plans for the plugin to be compatible with Craft 4.12+?

**Comment by marionnewlevant** [comment]
Finally merged in Brandon's changes, and Twig Perversion is working with Craft 5

**Comment by brettburwell** [comment]
+1  to what @anita-chouhan is saying. It'd be awesome to see this working in Craft 4 too—especially with the critical update that recently came out.

**Comment by brandonkelly** [comment]
My PR was merged into the `v2` branch (which is where 3.0.0 lives for Craft 4). So until it gets tagged, you can pull that branch into your projects by changing your `marionnewlevant/twig-perversion` requirement in composer.json to:

```json
"marionnewlevant/twig-perversion": "v2.x-dev as 3.0.0",
```

Then run `composer update`.

**Comment by brettburwell** [comment]
@brandonkelly Beautiful. I didn't realize that. Thanks so much for chiming in.
